### PR TITLE
Made sure an external id can be associated to an already-registered key.

### PR DESCRIPTION
**Pull request by xavierlepretre**
Before, in both implementations of `IdentityService`, when calling `registerKey` with an external id, and the key was already existing, then the external id would be ignored.

Now, I changed so that if the key is known but has no external id, then the external id is associated with it.

This is my 2nd PR.

# PR Checklist:

- [x] Have you run the unit, integration and smoke tests as described [here](https://docs.corda.net/head/testing.html)?
- [ ] If you added public APIs, did you write the JavaDocs?
- [ ] If the changes are of interest to application developers, have you added them to the [changelog](https://docs.corda.net/head/changelog.html) (`/docs/source/changelog.rst`), and potentially the [release notes](https://docs.corda.net/head/release-notes.html) (`/docs/source/release-notes.rst`)?
- [ ] If you are contributing for the first time, please read the [contributor agreement](https://docs.corda.net/head/contributing.html) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://docs.corda.net/head/contributing.html#merging-the-changes-back-into-corda).

Thanks for your code, it's appreciated! :)

### CODE REVIEW

**Comment by roastario** [comment]
thanks for this - it looks good to me, just going to have a talk with @roger3cev and a few others tomorrow before merging.

**Comment by roger-that-dev** [comment]
Thanks for the PR. What's the use-case where you need to assign an external ID to an already existing key? Surely when you generate a public key you know which account it's for, there and then?

**Comment by roger-that-dev** [comment]
Also, as I understand it, this is being changed for the training? Is that really necessary? @mnesbit - thoughts? This impacts kernel.

**Comment by alch3mistdev** [comment]
This is a particular scenario utilising the requested behaviour.

https://github.com/corda/corda-training-code/blob/3ad90b19b2bb0febc5d996b769cc9c63c3cbcb81/040-accounts-lib/workflows/src/test/java/com/template/account/AccountVaultCourseExercise.java#L99

@xavierlepretre feel free to comment or add any context if needed.

**Comment by xavierlepretre** [comment]
The use case pasted above is in a unit test demonstrating the association, or absence thereof, between an id and a pub key. 
The post hoc setting of an id to the pub key is for the benefit of a node that did not create the key in the first place but instead received it by one of many ways. Say:

- Alice creates key1 and key2 with account of id1
- Alice launches a flow that uses `SyncKeyMappingFlow` to send key1 over to Bob. It works. There is no id1 associated to key1 because this is not what `SyncKeyMappingFlow` does.
- Later Bob launches a flow that use `SyncKeyMappingFlow` to send key1 over to Carol. It works. There is no id1 associated...
- Later, Alice launches another flow that uses `ShareStateAndSyncAccountsFlow` to send the account id1 info plus key1 and key2 over to Carol. Carol is surprised that her vault now has 2 pub keys supposedly from the account Alice sent, but only key2 has the corresponding id1.

What do we tell Carol? Sorry mate, Alice should not have launched a flow with Bob that used `SyncKeyMappingFlow`. Better luck next time.

This PR makes sure that key1 is also associated with id1 whatever the previous choices of Alice.

**Comment by roger-that-dev** [comment]
OK, I get your point. I'll leave it up to @mnesbit as it's his call. Cheers

**Comment by alch3mistdev** [comment]
@rekalov Could you please take a look at this when you have a moment - to make sure it's compatible with any refactoring you're doing? thx

**Comment by rekalov** [comment]
I doesn't conflict with cert rotation work which I'm doing at the moment. Just check that changes are still mergeable after https://github.com/corda/corda/pull/6285.

As for the change itself, I don't know much about implementation, so can't comment if it should/can be changed in the node or in the flows.

One side note: it's not a thread safe, in the sense that two concurrent `registerKey` updating `externalId` may result in unpredictable behavior. Though, existing implementation of `registerKey` may also have such issues.

File: node/src/main/kotlin/net/corda/node/services/identity/InMemoryIdentityService.kt [hunk 1]
```diff
@@ -131,13 +131,19 @@ class InMemoryIdentityService(
...
         } else {
             if (party.name != existingEntry) {
             }
         }
+        if (externalId != null) {
+            val existingExtId = keyToExternalId[publicKeyHash]
+            if (existingExtId == null) {
+                registerKeyToExternalId(publicKey, externalId)
+            } else {
+                if (existingExtId != externalId) {
```
Comment (mnesbit) [inline node/src/main/kotlin/net/corda/node/services/identity/InMemoryIdentityService.kt]: what is meant to go in this branch of if tree?

Comment (roger-that-dev) [reply node/src/main/kotlin/net/corda/node/services/identity/InMemoryIdentityService.kt]: nothing, probably a bad merge

**Comment by roger-that-dev** [comment]
@rekalov with the existing implementation you couldn't update a key which was already registered. Instead you could only ever add a key and assign it to an external ID.

**Comment by rekalov** [comment]
@roger3cev , are you saying that it's not possible to have two concurrent `registerKey` calls on the same  key?

File: node/src/main/kotlin/net/corda/node/services/identity/InMemoryIdentityService.kt (hunk 1 above)
Comment (xavierlepretre) [reply node/src/main/kotlin/net/corda/node/services/identity/InMemoryIdentityService.kt]: It mimics the absence of a throw like on lines 135-136 of this very file.
The same `if` structures, but with throws, are [found](https://github.com/corda/corda/pull/6256/files/776e27b537f895417a28700cd1cb05b6f46ba078#diff-4242a6f0ed90c990808e6d0126ce9281R402-R416) in the `PersistentIdentityService` file.

Comment (mnesbit) [reply node/src/main/kotlin/net/corda/node/services/identity/InMemoryIdentityService.kt]: our coding standard is no dead, or commented out code so please remove empty if statement

Comment (mnesbit) [reply node/src/main/kotlin/net/corda/node/services/identity/InMemoryIdentityService.kt]: remove empty logic

**Comment by roger-that-dev** [comment]
@rekalov In the original version, you could only ever create a key and assign it to an ID in the same database transaction. So, from from the perspective of another flow... the key either exists and is maybe assigned to a UUID, or it doesn't exist. But a key that is not assigned to an ID, cannot be assigned one after it was created as the api didn't support it. So... even though it's highly unlikely two flows would choose to concurrently register the same key at the same time, even if they did there would be no strange side effects. The database would serialise the transactions and the second call to register key would be idempotent. I haven't looked at the new code, though.

edit: the above assumes flows don't run in parallel. If they do run in parallel, two flows looking to register the same key shouldn't cause any conflict, assuming they are registering it to the same party, which they should be!

**Comment by rekalov** [comment]
Even registering the same key (against the same party) may result in duplicated DB calls. 
It's fully OK for `keyToName` and `hashToKey` which are implemented using `AppendOnlyPersistentMap` and should be safe.

However, it's different for `registerKeyToExternalId()` which uses pure Hibernate `session.persist()` under the hood. Most likely, the second call will fail.

File: node/src/main/kotlin/net/corda/node/services/identity/InMemoryIdentityService.kt (hunk 1 above)
Comment (xavierlepretre) [reply node/src/main/kotlin/net/corda/node/services/identity/InMemoryIdentityService.kt]: Should I also remove the empty block on lines 135-136?

**Comment by roger-that-dev** [comment]
@rekalov ah that's true... even still... a _very_ unlikely scenario.

**Comment by rekalov** [comment]
Thanks @roger3cev , understood.

File: node/src/main/kotlin/net/corda/node/services/identity/InMemoryIdentityService.kt (hunk 1 above)
Comment (mnesbit) [reply node/src/main/kotlin/net/corda/node/services/identity/InMemoryIdentityService.kt]: yes just required logic blocks should be there. The others just confuse and risk merge problems at some point

**Comment by roger-that-dev** [comment]
@rekalov that doesn't mean we shouldn't fix it though :)

**Comment by nargas-ritu** [comment]
Internal backlog ticket [CORDA-4205](https://r3-cev.atlassian.net/browse/CORDA-4205)

**Comment by marcellofarina75** [comment]
Hello [xavierlepretre](https://github.com/xavierlepretre),
thank you for your contribution to Corda!

Our engineers reviewed your Pull Request and established that, while your proposal would definitely improve Corda, the work needed to merge it (in terms of unit, integration and regression tests, documentation updates, security assessments, and ongoing maintenance) would prevent us from shipping features and fixes that are prioritised higher in our roadmap (which is driven directly from inputs from across the user base).

So this means we regretfully need to let you know that your PR will be declined.

Please feel free to contact [devrel@r3.com](mailto:devrel@r3.com) if you need any further information

**Pull request closed without merging**